File: Example/CoreData/SyncKitCoreDataExample/QSEmployeeTableViewController.m
//
//  QSEmployeeTableViewController.m
//  SyncKit
//
//  Created by Manuel Entrena on 01/08/2016.
//  Copyright © 2016 Manuel. All rights reserved.
//

#import "QSEmployeeTableViewController.h"
#import "QSEmployee+CoreDataClass.h"

@interface QSEmployeeTableViewController () <NSFetchedResultsControllerDelegate, UIImagePickerControllerDelegate, UINavigationControllerDelegate>

@property (nonatomic, strong) NSFetchedResultsController *fetchedResultsController;

@property (nonatomic, strong) QSEmployee *editingEmployee;
@property (nonatomic, weak) IBOutlet UIButton *createButton;

@end

@implementation QSEmployeeTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupFetchedResultsController];
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)setupFetchedResultsController
{
    if (!_fetchedResultsController) {
        NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:@"QSEmployee"];
        fetchRequest.predicate = [NSPredicate predicateWithFormat:@"company == %@", self.company];
        fetchRequest.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES]];
        
        _fetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest managedObjectContext:self.managedObjectContext sectionNameKeyPath:nil cacheName:nil];
        _fetchedResultsController.delegate = self;
        [_fetchedResultsController performFetch:nil];
    }
}

- (void)createEmployeeWithName:(NSString *)name
{
    QSEmployee *employee = [NSEntityDescription insertNewObjectForEntityForName:@"QSEmployee" inManagedObjectContext:self.managedObjectContext];
    employee.name = name;
    employee.company = self.company;
    employee.identifier = [[NSUUID UUID] UUIDString];
    [self.managedObjectContext save:nil];
}

#pragma mark - NSFetchedResultsControllerDelegate

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView beginUpdates];
}


- (void)controller:(NSFetchedResultsController *)controller didChangeSection:(id <NSFetchedResultsSectionInfo>)sectionInfo
           atIndex:(NSUInteger)sectionIndex forChangeType:(NSFetchedResultsChangeType)type {
    
    switch(type) {
        case NSFetchedResultsChangeInsert:
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                          withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeDelete:
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                          withRowAnimation:UITableViewRowAnimationFade];
            break;
        default:
            break;
    }
}


- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject
       atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath {
    
    UITableView *tableView = self.tableView;
    
    switch(type) {
            
        case NSFetchedResultsChangeInsert:
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeDelete:
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeUpdate:
            [self configureCell:[tableView cellForRowAtIndexPath:indexPath]
                    atIndexPath:indexPath];
            break;
            
        case NSFetchedResultsChangeMove:
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
    }
}


- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView endUpdates];
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleDelete;
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [[self.fetchedResultsController sections] count];
}

- (NSInteger)tableView:(UITableView *)table numberOfRowsInSection:(NSInteger)section {
    if ([[self.fetchedResultsController sections] count] > 0) {
        id <NSFetchedResultsSectionInfo> sectionInfo = [[self.fetchedResultsController sections] objectAtIndex:section];
        return [sectionInfo numberOfObjects];
    } else
        return 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    [self configureCell:cell atIndexPath:indexPath];
    
    return cell;
}

- (void)configureCell:(UITableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath
{
    QSEmployee *employee = [self.fetchedResultsController objectAtIndexPath:indexPath];
    if (employee.photo) {
        cell.imageView.image = [[UIImage alloc] initWithData:employee.photo];
    } else {
        cell.imageView.image = nil;
    }
    cell.textLabel.text = employee.name ?: @"Object name is nil";
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    return self.canWrite;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        QSEmployee *employee = [self.fetchedResultsController objectAtIndexPath:indexPath];
        [self.managedObjectContext deleteObject:employee];
        [self.managedObjectContext save:nil];
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (!self.canWrite) {
        [self showReadOnlyPermission];
        return;
    }
    
    QSEmployee *employee = [self.fetchedResultsController objectAtIndexPath:indexPath];
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Update employee" message:nil preferredStyle:UIAlertControllerStyleAlert];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Add photo" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        
        [self presentImagePickerForEmployee:employee];
    }]];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Clear photo" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        employee.photo = nil;
        [self.managedObjectContext performBlock:^{
            [self.managedObjectContext save:nil];
        }];
    }]];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Clear name" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        employee.name = nil;
        [self.managedObjectContext performBlock:^{
            [self.managedObjectContext save:nil];
        }];
    }]];
    [alertController addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"Enter new name";
    }];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        employee.name = alertController.textFields.firstObject.text;
        [self.managedObjectContext performBlock:^{
            [self.managedObjectContext save:nil];
        }];
    }]];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    
    [self presentViewController:alertController animated:YES completion:nil];
}

#pragma mark - Image

- (void)presentImagePickerForEmployee:(QSEmployee *)employee
{
    self.editingEmployee = employee;
    UIImagePickerController *imagePickerController = [[UIImagePickerController alloc] init];
    imagePickerController.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    imagePickerController.delegate = self;
    [self presentViewController:imagePickerController animated:YES completion:nil];
}

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info
{
    UIImage *image = info[@"UIImagePickerControllerOriginalImage"];
    
    UIImage *resizedImage = [self imageWithImage:image scaledToSize:CGSizeMake(150, 150)];
    
    self.editingEmployee.photo = UIImagePNGRepresentation(resizedImage);
    
    [self.managedObjectContext performBlock:^{
        [self.managedObjectContext save:nil];
    }];
    
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (UIImage *)imageWithImage:(UIImage *)image scaledToSize:(CGSize)newSize
{
    UIGraphicsBeginImageContextWithOptions(newSize, NO, 0.0);
    [image drawInRect:CGRectMake(0, 0, newSize.width, newSize.height)];
    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return newImage;
}

- (void)showReadOnlyPermission
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Read only" message:@"You only have read permission for this employee" preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
    [self presentViewController:alertController animated:YES completion:nil];
}

#pragma mark - Actions

- (IBAction)insertNewEmployee:(id)sender
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"New employee" message:nil preferredStyle:UIAlertControllerStyleAlert];
    [alertController addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"Enter employee's name";
    }];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Add" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self createEmployeeWithName:alertController.textFields.firstObject.text];
    }]];
    [self presentViewController:alertController animated:YES completion:nil];
}

@end

File: Example/CoreData/SyncKit/QSEmployeeTableViewController.swift
//
//  QSEmployeeSwiftTableViewController.swift
//  SyncKitCoreDataExample
//
//  Created by Jérôme Haegeli on 25.07.18.
//  Copyright © 2018 Manuel. All rights reserved.
//

import UIKit
import CoreData
import UIKit

class QSEmployeeTableViewController: UITableViewController, NSFetchedResultsControllerDelegate, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    var managedObjectContext: NSManagedObjectContext?
    var company: QSCompany?
    var canWrite = false
    
    private var fetchedResultsController: NSFetchedResultsController<NSFetchRequestResult>?
    private var editingEmployee: QSEmployee?
    @IBOutlet private weak var createButton: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupFetchedResultsController()
        tableView.reloadData()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        createButton.isHidden = !canWrite
    }
    
    func setupFetchedResultsController() {
        if !(fetchedResultsController != nil) {
            let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "QSEmployee")
            fetchRequest.predicate = NSPredicate(format: "company == %@", company!)
            fetchRequest.sortDescriptors = [NSSortDescriptor(key: "name", ascending: true)]
            fetchedResultsController = NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: managedObjectContext!, sectionNameKeyPath: nil, cacheName: nil)
            fetchedResultsController?.delegate = self
            try? fetchedResultsController?.performFetch()
        }
    }
    
    func createEmployee(withName name: String?) {
        let employee = NSEntityDescription.insertNewObject(forEntityName: "QSEmployee", into: managedObjectContext!) as? QSEmployee
        employee?.name = name
        employee?.company = company
        employee?.identifier = UUID().uuidString
        do {
            try self.managedObjectContext?.save()
        } catch {
            print(error)
        }
    }

//pragma mark - NSFetchedResultsControllerDelegate

    func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.beginUpdates()
    }

    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange sectionInfo: NSFetchedResultsSectionInfo, atSectionIndex sectionIndex: Int, for type: NSFetchedResultsChangeType) {
        switch type {
        case .insert:
            tableView.insertSections(NSIndexSet(index: sectionIndex) as IndexSet, with: .fade)
        case .delete:
            tableView.deleteSections(NSIndexSet(index: sectionIndex) as IndexSet, with: .fade)
        default:
            break
        }
    }

    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        let tableView: UITableView? = self.tableView
        switch type {
        case .insert:
            tableView?.insertRows(at: [newIndexPath!], with: .fade)
        case .delete:
            tableView?.deleteRows(at: [indexPath!], with: .fade)
        case .update:
            if let aPath = indexPath {
                configureCell(tableView?.cellForRow(at: aPath), at: indexPath)
            }
        case .move:
            tableView?.deleteRows(at: [indexPath!], with: .fade)
            tableView?.insertRows(at: [newIndexPath!], with: .fade)
        }
    }

    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.endUpdates()
    }
    
    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        return .delete
    }

//pragma mark - Table view data source
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        if let sections = fetchedResultsController?.sections {
            return sections.count
        }
        return 0
    }
    
    override func tableView(_ table: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        if let sections = fetchedResultsController?.sections {
            let currentSection = sections[section]
            return currentSection.numberOfObjects
        }
        
        return 0
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: "cell")
        configureCell(cell, at: indexPath)
        if let aCell = cell {
            return aCell
        }
        return UITableViewCell()
    }
    
    func configureCell(_ cell: UITableViewCell?, at indexPath: IndexPath?) {
        var employee: QSEmployee? = nil
        if let aPath = indexPath {
            employee = fetchedResultsController?.object(at: aPath) as? QSEmployee
        }
        if employee?.photo != nil {
            if let aPhoto = employee?.photo {
                cell?.imageView?.image = UIImage(data: aPhoto as Data)
            }
        } else {
            cell?.imageView?.image = nil
        }
        cell?.textLabel?.text = employee?.name ?? "Object name is nil"
    }

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return canWrite
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            let employee = fetchedResultsController?.object(at: indexPath) as? QSEmployee
            if let anEmployee = employee {
                managedObjectContext?.delete(anEmployee)
            }
            try? managedObjectContext?.save()
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if !canWrite {
            showReadOnlyPermission()
            return
        }
        let employee = fetchedResultsController?.object(at: indexPath) as? QSEmployee
        let alertController = UIAlertController(title: "Update employee", message: nil, preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "Add photo", style: .default, handler: { action in
            self.presentImagePicker(for: employee)
        
        }))
        alertController.addAction(UIAlertAction(title: "Clear photo", style: .default, handler: { action in
            employee?.photo = nil
            self.managedObjectContext?.perform({
                try? self.managedObjectContext?.save()
            })
        }))
        alertController.addAction(UIAlertAction(title: "Clear name", style: .default, handler: { action in
            employee?.name = nil
            self.managedObjectContext?.perform({
                try? self.managedObjectContext?.save()
            })
        }))
        alertController.addTextField(configurationHandler: { textField in
            textField.placeholder = "Enter new name"
        })
        alertController.addAction(UIAlertAction(title: "Ok", style: .default, handler: { action in
            employee?.name = alertController.textFields?.first?.text
            self.managedObjectContext?.perform({
                try? self.managedObjectContext?.save()
            })
        }))
        alertController.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        present(alertController, animated: true)
    }

    // MARK: - Image
    func presentImagePicker(for employee: QSEmployee?) {
        editingEmployee = employee
        let imagePickerController = UIImagePickerController()
        imagePickerController.sourceType = .photoLibrary
        imagePickerController.delegate = self
        present(imagePickerController, animated: true)
    }
    
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        let image = info["UIImagePickerControllerOriginalImage"] as? UIImage
        let resizedImage: UIImage? = self.image(with: image, scaledTo: CGSize(width: 150, height: 150))
        editingEmployee?.photo = UIImagePNGRepresentation(resizedImage!) as! NSData
        managedObjectContext?.perform({
            try? self.managedObjectContext?.save()
        })
        dismiss(animated: true)
    }
    
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        dismiss(animated: true)
    }
    
    func image(with image: UIImage?, scaledTo newSize: CGSize) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(newSize, false, 0.0)
        image?.draw(in: CGRect(x: 0, y: 0, width: newSize.width, height: newSize.height))
        let newImage: UIImage? = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage
    }
    
    func showReadOnlyPermission() {
        let alertController = UIAlertController(title: "Read only", message: "You only have read permission for this employee", preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        present(alertController, animated: true)
    }
    
    // MARK: - Actions
    @IBAction func insertNewEmployee(_ sender: Any) {
        let alertController = UIAlertController(title: "New employee", message: nil, preferredStyle: .alert)
        alertController.addTextField(configurationHandler: { textField in
            textField.placeholder = "Enter employee's name"
        })
        alertController.addAction(UIAlertAction(title: "Add", style: .default, handler: { action in
            self.createEmployee(withName: alertController.textFields?.first?.text)
        }))
        present(alertController, animated: true)
    }

    
}


